bestie/server/service.go

```diff
@@ -97,3 +139,38 @@ func (s *serviceStats) incrementEventsTotal(bevid interface{}) {
 	}
 	s.eventsTotal.WithLabelValues(label).Inc()
 }
+
+// BigQuery operational exceptions.
+func (s *serviceStats) incrementBigQueryExcessDelay() {
+	s.bigQueryExceptionTotal.WithLabelValues("excess_delay").Inc()
+}
+func (s *serviceStats) incrementBigQueryProtobufError() {
+	s.bigQueryExceptionTotal.WithLabelValues("protobuf_error").Inc()
+}
+func (s *serviceStats) incrementBigQueryTableNotFound() {
+	s.bigQueryExceptionTotal.WithLabelValues("table_not_found").Inc()
+}
+
+// BigQuery row insertion delay (histogram).
+func (s *serviceStats) incrementBigQueryInsertDelay(seconds int) {
+	s.bigQueryInsertDelay.Observe(float64(seconds))
+}
+
+// BigQuery row insertion outcome.
+func (s *serviceStats) incrementBigQueryInsertError() {
+	s.bigQueryInsertTotal.WithLabelValues("error").Inc()
+}
+func (s *serviceStats) incrementBigQueryInsertOK() {
+	s.bigQueryInsertTotal.WithLabelValues("ok").Inc()
+}
+func (s *serviceStats) incrementBigQueryInsertTimeout() {
+	s.bigQueryInsertTotal.WithLabelValues("timeout").Inc()
+}
+
+// BigQuery total metrics processed.
+func (s *serviceStats) incrementBigQueryMetricDiscard(count int) {
+	s.bigQueryMetricTotal.WithLabelValues("discard").Add(float64(count))
+}
+func (s *serviceStats) incrementBigQueryMetricUpload(count int) {
+	s.bigQueryMetricTotal.WithLabelValues("upload").Add(float64(count))
+}
```

Comment:
Can all this be inlined?

Reply:
GitHub seems to have lost my first attempt at replying, so I'll try again...

The intent of these methods is to abstract away the details of Prometheus metrics from the rest of the code. I've completely changed my approach, replacing these individual methods with a few general purpose functions that increment the service counters based on counter IDs. These new functions conveniently wrap a call to a common update function that is invoked as a goroutine, thus addressing Adam's comment on this subject.